node/rpc/src/state_api.rs

```diff
@@ -186,3 +190,91 @@ pub(crate) async fn state_wait_msg<DB: Blockstore + Clone + Send + Sync + 'stati
+//
+/// Traverse an IPLD directed acyclic graph and use libp2p-bitswap to request any missing nodes.
+/// This function has two primary uses: (1) Downloading specific state-roots when Forest deviates
+/// from the mainline blockchain, (2) fetching historical state-trees to verify past versions of the
+/// consensus rules.
+pub(crate) async fn state_fetch_root<DB: Blockstore + Clone + Sync + Send + 'static, B: Beacon>(
+    data: Data<RPCState<DB, B>>,
+    Params((CidJson(root_cid),)): Params<StateFetchRootParams>,
+) -> Result<StateFetchRootResult, JsonRpcError> {
+    const MAX_CONCURRENT_REQUESTS: usize = 16;
+    const REQUEST_TIMEOUT: Duration = Duration::from_secs(10);
+
+    let sem = Arc::new(Semaphore::new(MAX_CONCURRENT_REQUESTS));
+    let mut seen: HashSet<Cid> = HashSet::new();
+    let mut counter: usize = 0;
+    let mut failures: usize = 0;
+    let mut task_set = JoinSet::new();
+
+    let mut get_ipld_link = |ipld: &Ipld| match ipld {
+        Ipld::Link(cid) if cid.codec() == 0x71 && seen.insert(*cid) => Some(*cid),
+        _ => None,
+    };
+
+    task_set.spawn(async move { Ok(Ipld::Link(root_cid)) });
+
+    // Iterate until there are no more ipld nodes to traverse
+    while let Some(result) = task_set.join_next().await {
+        match result? {
+            Ok(ipld) => {
+                for new_cid in ipld.iter().filter_map(&mut get_ipld_link) {
+                    counter += 1;
+                    if counter % 1_000 == 0 {
+                        // set RUST_LOG=forest_rpc::state_api=debug to enable these printouts.
+                        log::debug!(
+                            "Still downloading. Fetched: {counter}, Failures: {failures}, Concurrent: {}",
+                            MAX_CONCURRENT_REQUESTS - sem.available_permits()
+                        );
+                    }
+                    task_set.spawn({
+                        let network_send = data.network_send.clone();
+                        let db = data.chain_store.db.clone();
+                        let sem = sem.clone();
+                        async move {
+                            if !db.has(&new_cid)? {
+                                // If a CID isn't in our database, request it via bitswap (limited
+                                // by MAX_CONCURRENT_REQUESTS)
+                                let permit = sem.acquire_owned().await?;
+                                let (tx, rx) = flume::bounded(1);
+                                network_send
+                                    .send_async(NetworkMessage::BitswapRequest {
+                                        epoch: 0,
+                                        cid: new_cid,
+                                        response_channel: tx,
+                                    })
+                                    .await?;
+                                // Bitswap requests do not fail. They are just ignored if no-one has
```

Comment:
FYI there is a `send_dont_have` option in `BitswapRequest` to ask peers to respond even if they don't have the requested block, if that could be beneficial here.